plugins/file-open-save-new/core/src/main/java/org/pentaho/di/plugins/fileopensave/dragdrop/ElementTreeDropAdapter.java

```diff
@@ -60,8 +60,8 @@ public ElementTreeDropAdapter( TreeViewer viewer, LogChannelInterface log ) {
   public boolean performDrop( Object data ) {
     int location = getCurrentLocation();
     Object genericTarget = getCurrentTarget() == null ? getViewer().getInput() : getCurrentTarget();
-    if ( !( genericTarget instanceof Entity ) || ( genericTarget instanceof VFSLocation
-      && ( (VFSLocation) genericTarget ).hasBuckets() ) ) {
+    if ( !( genericTarget instanceof Entity )
+      || ( genericTarget instanceof VFSDirectory && !( (VFSDirectory) genericTarget ).isCanAddChildren() ) ) {
```

Comment:
Removed `VFSLocation#hasBuckets()` whose only use was here. At the same time, fixing a bug where it would be possible to drop a file in a bucketed connection root folder. Will update this comment when JIRA issue is created.

Reply:
From what I remember.  It should be possible to drop a file on the connection name of a HadoopCluster connection in order to copy something to the root folder of hdfs.  This cannot be allowed in buckets, however, which the file browser cannot create.  This is why the logic is there.  To allow copies to the root folder of hdfs while preventing users from doing the same to "bucketed" VFS connection which, if possible, would create a new bucket.

I believe your change here broke this as copying to the any VFSLocation is no longer prohibited.

Reply:
First of all, please note that `VFSLocation` extends `VFSDirectory`.

There was a bug in which the `isCanAddChildren()` property of `VFSLocation` objects was not being set to the right value. The right value should reflect whether the corresponding connection has buckets (actually, if it is using buckets, given the root path changes this). This bug was causing the File Open Save dialog's "Add folder" button to not be enabled/disabled correctly for connection root folders (it was exactly opposite). The same condition that controls whether a child folder can be created in a parent folder can/should be used to control whether a file or folder can be dropped into a parent folder.

Having fixed the original bug (fix: https://github.com/pentaho/pentaho-kettle/pull/9334/files#diff-2f6c36b8d755e99ea63e13facad918e197b7d844cfb00c0f3da118fd363f3a38R182) that affected all connection types, and not only S3 and SNW, it is now possible to read a directory's `isCanAddChildren` to validate if a drop can happen.

Change to `VFSLocation`: https://github.com/pentaho/pentaho-kettle/pull/9334/files#r1624277097

I will be creating a bug issue and will update this comment with it.